### ISSUE TITLE
ui: GiB vs. GB (1024^3 vs. 1000^3)

### ISSUE THREAD

**Issue by spencerkimball** [cockroachdb/cockroach#4735]
tl;dr Let's allow the *iB designations, but make the *B designations also use 1024^i and also make them the defaults for documentation and user interface output.

Let's not confuse our users the way that hard drive manufacturers confused theirs in the 90s. I'm fine with allowing the [MGTPE]iB designations, but I strongly disagree with making [MBTPE]B mean 1000^i bytes instead of 1024^i bytes. Most people expect 1024^i, so that's what we should use. If you Google the difference, you'll find _many_ explanations. Most people are very confused when they see the "i". I didn't know what it meant, for example. I'm sure others here didn't know. My expectation is that byte sizes are specified in 1024^i increments. Why are we actually agreeing with the idiocy of hard drive manufacturers and adopting their nomenclature?

Other tools I've used which specify byte sizes in 1024^i increments using the standard [MGTPE]B designations:

```
pprof
colossus
leveldb/rocksdb
hdfs
cassandra
java
```

cc @BramGruneir

**Comment by mrtracy** [comment]
Normal metric prefixes (without the 'i') should always denote 1000^x units, according to both IEC and SI.  The usage of normal metric prefixes to denote 1024^x units is the incorrect notation; that's actually the reason that the 'i' prefixes were created in the first place, since the new 'i' prefixes were unambiguous in their meaning.

Note that MacOSX and Ubuntu Linux report file sizes in regular metric prefixes, and they do mean the correct 1000^i. In fact, this is a relatively recent development; wikipedia says that MacOSX had the incorrect behavior (i.e. "Megabyte" = 1024^3) until OSX 1.6, when it was changed.

So, if you want to report units in normal metric prefixes such as "MB", that is fine - but they _must_ mean 1000^x units, or we are perpetuating the problem.

**Comment by spencerkimball** [comment]
The issue is that MiB is not familiar to many users. You shouldn't have to know what a "mebibyte" or "gibibyte" is to view the admin UI. It certainly _is_ a goal for our users to see units in the dashboard  without being confused as to what they mean.

To that end, why don't we compromise and allow [MGTPE] as suffix units, meaning 1024^3, and also use those as the default units in our user interfaces. MB and GB can continue as 10^6 and 10^9 in the small set of use cases where operators might choose to input sizes using those units via the command line.

**Comment by mrtracy** [comment]
Spencer and I discussed this a bit offline.

Owing to the tepid response received from using the binary prefixes ([MGTPE]i), we would prefer to use the normal metric prefixes (sans the "i") unless ambiguity around those prefixes was truly extreme. Although there is some ambiguity, we believe it is surmountable with clear documentation.

Additionally, even though the standards-body definition of "Megabyte" is now 1000^3 bytes, that is largely irrelevant; there is already so much ambiguous usage of this unit, that Cockroach _must_ provide discoverable documentation as to what it means, regardless of the decision.

Therefore, our current recommended solution:

+ Display bytes units with [MGTPE] prefixes everywhere in the UI, where each unit represents a power of + Provide _discoverable_ documentation which describes Cockroaches specific choice to do this (as opposed to [MGTPE] representing powers of 10).

A final note: compromise in which plain [MGTPE] prefixes mean different values in different contexts is not acceptable.  If we go with go with powers of 2, it must be consistent everywhere.

**Comment by RaduBerinde** [comment]
I agree.

Personally I am baffled whenever Megabyte does NOT mean 1024^2 bytes. I understand SI but really the 1024 is so ingrained and natural when working with computers that it makes no sense to try to switch it. Mebibyte/Gibibyte frankly sound ridiculous so no wonder nobody uses them.

FWIW commands like `du`, `df` on my Debian 8 use 1024 (they do support an `--si` flag but it's not the default)

**Comment by bdarnell** [comment]
I don't care enough to fight the powers-of-2 crowd on this, but I'll just say that I much prefer the powers of 10. I'll often do things like `du -k` when I'm expecting files in megabytes and lop off three digits in my head. Yeah, this is nonsense because I do not also pass `--si` so I'm mixing the two bases (Like I said, I don't care too much about this issue). But now that we're not fighting for every bit of address space, powers of 10 are just more convenient for humans.

**Comment by johnmuth81** [comment]
> On Feb 29, 2016, at 2:49 PM, Spencer Kimball <notifications@github.com> wrote:
> 
> tl;dr Let's allow the *iB designations, but make the *B designations also use 1024^i and also make them the defaults for documentation and user interface output.
> 
> Let's not confuse our users the way that hard drive manufacturers confused theirs in the 90s. I'm fine with allowing the [MGTPE]iB designations, but I strongly disagree with making [MBTPE]B mean 1000^i bytes instead of 1024^i bytes.
> 
Hi Spencer,

If you go into the enterprise world, you will confuse people with MB=1024^2. People in the enterprise space know and expect B vs iB as defined by IEC, since they are they ones who insisted drive (and array) mfg's disambiguate the terms in the first place.

Yeah it’s stupid, but not going along with it will violate the principle of least surprise in the enterprise space.

Regards,
John Muth

**Comment by knz** [comment]
I think it's fair for CockroachDB to want to use the power-of-two encoding in the UI by default, but then when *displaying* it should do so with the SI units (with the 'i'). When reading people will either know what it means or just discard it mentally and still get the right picture.

When accepting input (like e.g. for `--store`) I think it's acceptable to only accept the power of two encoding regardless of whether the 'i' was specified, *but*:

- it needs to be properly documented
- I would personally value an informational message when the user enters the value without 'i' to remind them that we are adding it implicitly, in case the user is one of those who knows the difference.

**Comment by JackKrupansky** [comment]
Powers of 2?

Geez, if you guys keep this up CockroachDB will quickly get a reputation as being "A database by engineers, for engineers." Is that what you want??

(awaiting snarky response)

GiB seems like a term invented by a product manager to appeal to other product managers. ("A database by product managers, for product managers" - seems the norm, but just doesn't sound right to me.)

It feels like a fielder's choice to me. Either is workable. Yeah, I suppose I am partial to powers of 2, but I grew up in an age when a computer the size of a room had 64K of memory (I'd write 64KiB but it was actually 64 *1024 36-bit "words".)

Personally, when I see a number expressed as 60 MB I explicitly make no assumptions as to whether the writer meant 60 * 1024 ^ 2 or 60 million and that that's okay since it is frequently only an approximation anyway. But when I see 64 MB I presume that they are hinting that powers of 2 are involved. Ditto for 200 GB vs. 256 GB.

Hmmm... remind me... the default limit of 64 MB for size of a range means... 64 * 1024 ^ 2 or 64 million?! The design spec says 64 M, leaving M ambiguous as to MiB or million. (Okay, I peeked at config.go and see that it's... 67108864 which is 64 MiB.)

**Comment by spencerkimball** [comment]
@johnmuth81 thanks for that perspective. I think for the next year or so we'll be more focused on a developer audience than an enterprise one.

@knz I was unaware of what MiB meant when I saw it. I actually did not discard it mentally and assume it meant the same thing as 1024^2. I thought we might be measuring something else, like mega bits because of the lowercase "i". I was unaware that mebibytes existed (btw, it's not even recognized by spell check) until I saw them. I get that I must have seen them somewhere at some point but it hadn't ever sunk in over the course of decades.

I would argue strongly that principal of least surprise is to use MB and make it mean 1024^2. Exactly to your original argument, when people see MB who are aware of the distinction between MiB and MB, they'll have to mouse over it to see what we actually mean, but I'd wager this represents less total surprise than introducing a sizable fraction of our audience to MiB for the first time.

**Comment by knz** [comment]
I'm not comfortable with being inaccurate for the convenience of uneducated users. Good software also makes a good impression when it educates.

Here's a compromise. When printing values use "K" "M" "G" instead of KB/KiB etc. This way uneducated users still understand and educated users do not get the wrong message. 

There's another advantage to avoiding the B and that's when printing throughputs. In network tech capital B often means bits not bytes. Without the B people default to bytes. That avoid having to clarify bit vs byte.

What do you think?
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by petermattis** [comment]
The whole `MB` vs `MiB` thing is a mess. This is quite confusing: `About This Mac` reports my laptop has `16 GB` of RAM and `500 GB` (500,277,790,720 bytes) of disk. Looks like Apple is using powers of 1024 for memory and powers of 1000 for disk.

**Comment by jseldess** [comment]
@spencerkimball, @petermattis, @bdarnell, reading through the above, I'm not sure if we've come to a decision yet. Assigning back to @spencerkimball for that. I'll update the docs once the change has been made.

**Comment by mberhault** [comment]
it doesn't seem as if we have either. However, just to throw my two cents of a Euro in:
* sizes in bytes should always be in powers of two. we're not a hard drive vendor. 
* sizes in units of something other than bytes (eg: counts of files, ranges, etc..) should be plain 1000 multiples
* GB vs GiB is just a never-ending argument. I will never understand GiB, but it could be due to my experience. However, as long as we stick to the right multiple based on the unit (^2 for bytes, ^10 for everything else), I don't care all that much, but I'll still flinch every time I see GiB.

**Comment by BramGruneir** [comment]
As most of you know, I'm 100% behind using the IEC units.  
There is no confusion there and if someone has to look it up, then they will, once.  

https://wiki.ubuntu.com/UnitsPolicy

On OSX, try `df -h`

**Comment by petermattis** [comment]
Perusing the kubernetes documentation this weekend I noticed that they have adopted the IEC units: http://kubernetes.io/docs/admin/resourcequota/. We might be on the wrong side of history if we insist on using MB/GB as powers-of-2.

**Comment by JackKrupansky** [comment]
@petermattis, are you more concerned about the need to use the "i" in GiB (and that GB uses 1,000 units) or whether the basic unit is 1024 vs 1000? For example, should range size be a multiple of 1024*1024 or 1,000,000?

To me, the more interesting question is not how to measure total size, which is usually just a rough approximation anyway (unless given in actual bytes), but the unit of transfer and unit of allocation. For example, what data block sizes are managed most efficiently by the operating system, I/O drivers, etc. If the answer is that RAM allocation and I/O transfers are sti

**Comment by spellgen** [comment]
Kilo, mega, giga are standard SI prefixes with very clear and well established meaning in science as specific powers of ten. There *are* cases when it makes sense to use those for bytes: data collected at a network interface, size of a file - not all measurements of bytes are inherently natural in powers-of-two.

When it comes to measuring things that are actually based on powers-of-two it is nice to be able to reach for something that gives you round numbers: 4 GiB. The difference is a little bit technical and academic but it is nice to be precise. However, how do I pronounce that? Four gigabytes.

**Comment by spencerkimball** [comment]
Looks like this is just going to stay GiB, MiB, etc. Inertia wins.

**Comment by tamird** [comment]
Not sure this is a case of inertia, but I'm happy with the outcome.
